pkg/services/collector.go

```diff
@@ -0,0 +1,512 @@
+
+	err := filepath.Walk(chartPath,
+		func(path string, info os.FileInfo, err error) error {
+			if err != nil {
+				return err
+			}
+			pathIsDirectory, err := isDirectory(path)
+			if err != nil {
+				return err
+			}
+			if !pathIsDirectory {
+				packageFiles = append(packageFiles, path)
+			}
+
+			return nil
+		})
+	if err != nil {
+		return err
+	}
+
+	out, err := os.Create(fmt.Sprintf("%s-%s.tgz", chartName, chartVersion))
+	if err != nil {
+		return err
+	}
+	defer out.Close()
+
+	// Create the archive and write the output to the "out" Writer
+	err = createArchive(packageFiles, fmt.Sprintf("%s/%s/", filepath.Dir(chartPath), filepath.Base(chartPath)), fmt.Sprintf("%s/", chartName), out)
+	return err
+}
+
+func (c *Collector) AddHelmRepositories(releaseContainer *gabs.Container) error {
+	if !releaseContainer.ExistsP("chart.metadata.dependencies") {
+		return nil
+	}
+	dependencies, err := releaseContainer.Path("chart.metadata.dependencies").Children()
+	if err != nil {
+		return err
+	}
+
+	var repositoryNames []string
+
+	for _, dependency := range dependencies {
+		dependencyMap, err := dependency.ChildrenMap()
+		if err != nil {
+			return err
+		}
+
+		repositoryName := dependencyMap["name"].Data().(string)
+		repositoryUrl := dependencyMap["repository"].Data().(string)
+
+		_, err = execCommand(fmt.Sprintf("helm repo add %s %s", repositoryName, repositoryUrl))
+		if err != nil {
+			return err
+		}
+
+		repositoryNames = append(repositoryNames, repositoryName)
+	}
+
+	_, err = execCommand(fmt.Sprintf("helm repo update %s", strings.Join(repositoryNames, " ")))
+	return err
+}
+
+func (c *Collector) UploadHelmPackage(chartName string, chartVersion string) error {
+	file, err := os.Open(fmt.Sprintf("%s-%s.tgz", chartName, chartVersion))
+	if err != nil {
+		return err
+	}
+	fileContents, err := ioutil.ReadAll(file)
+	if err != nil {
+		return err
+	}
+	fi, err := file.Stat()
+	if err != nil {
+		return err
+	}
+	file.Close()
+
+	body := new(bytes.Buffer)
+	writer := multipart.NewWriter(body)
+	part, err := writer.CreateFormFile("chart", fi.Name())
+	if err != nil {
+		return err
+	}
+	part.Write(fileContents)
+	writer.Close()
+	req, err := http.NewRequest("POST", fmt.Sprintf("%s/api/charts", c.ChartmuseumUrl), body)
+	if err != nil {
+		return err
+	}
+	req.SetBasicAuth(c.ChartmuseumUsername, c.ChartmuseumPassword)
+	req.Header.Add("Content-Type", writer.FormDataContentType())
+	resp, err := c.HttpClient.Do(req)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+	responseBody, err := io.ReadAll(resp.Body)
+	if resp.StatusCode != 201 {
+		return errors.New(fmt.Sprintf("Receiving list of charts failed. Status code - %d, Body - %s", resp.StatusCode, string(responseBody)))
+	}
+	c.ChartVersionsRegistry[chartName] = append(c.ChartVersionsRegistry[chartName], chartVersion)
+	fmt.Printf("Chart %s-%s has been saved successfully\n", chartName, chartVersion)
+	return nil
+}
+
+func (c *Collector) CheckAllSecrets() {
```

Comment:
This function is reeeeeeeallllly long, mb it can be split?

Reply:
agree, rewrote it